### PR TITLE
add/remove/git commits to gitlab using a gitlab token

### PR DESCRIPTION
**Pull request by J-Thompson12**
References: #724 

**What changed?**
Allow user to add/remove/get commits for a  gitlab user project and group project with gitlab token set. There was also changes made around normalizing a url. This part of it will need a good review as i removed sanitizeRepoUrl and that was used in a few random places. A repo can also be added that is part of a gitlab subgroup. There are not tests around subgroup and having a subgroup in a subgroup isnt supported.

Work that will be done in future prs:
- Add acceptance test with merge pr
- Add tests for subgroup
- Add ability to add a repo that is a subgroup in a subgroup

**Why?**

**How did you test it?**
unit tests and acceptance tests

**Release notes**
Add gitlab group project and user project if gitlab token is set as GITLAB_TOKEN

**Documentation Changes**
https://github.com/weaveworks/weave-gitops-docs/pull/83

### CODE REVIEW

File: test/acceptance/test/utils.go [hunk 1]
```diff
@@ -270,15 +279,54 @@ func createGitRepoBranch(repoAbsolutePath string, branchName string) string {
...
+			Domain:       gitlab.DefaultDomain,
+			Organization: GITLAB_ORG,
+		},
+	}
+
+	orgInfo, err := gitProvider.OrgRepositories().Get(context.Background(), orgRef)
+	Expect(err).ShouldNot(HaveOccurred())
+
+	visibility := string(*orgInfo.Get().Visibility)
+	log.Infof("Repo visibility private=%s", visibility)
```
Comment (jpellizzari) [inline test/acceptance/test/utils.go:327]: It's weird to me that we log things here. I'm assuming that we log things elsewhere in acceptance tests?

File: pkg/apputils/utils.go [hunk 2]
```diff
@@ -174,7 +174,12 @@ func getGitClients(ctx context.Context, url, configUrl, namespace string, isHelm
 }
 
 func getAuthService(ctx context.Context, providerUrl string) (auth.AuthService, error) {
-	gitProvider, providerErr := auth.GetGitProvider(ctx, providerUrl)
+	normalizedUrl, err := gitproviders.NewNormalizedRepoURL(providerUrl)
+	if err != nil {
+		return nil, fmt.Errorf("error normalizing url: %w", err)
+	}
+
+	gitProvider, providerErr := auth.GetGitProvider(ctx, normalizedUrl.String())
```
Comment (jpellizzari) [inline pkg/apputils/utils.go:182]: Might be an opportunity to change the `GetGitProvider` to accept a `NormalizedRepoURL`. It looks like that function is doing more work than it has to:

https://github.com/weaveworks/weave-gitops/blob/ae571fbc056cb5cd341a9f99b029b7f6f92c6258/pkg/services/auth/auth.go#L49

**Review by jpellizzari — COMMENTED** [review]
Looks pretty good. There are some opportunities for refactoring and type safety that we might be able to take advantage of. Let me know if you want to hop on a call.

File: pkg/gitproviders/provider.go [hunk 3]
```diff
@@ -254,7 +254,7 @@ func (p defaultGitProvider) GetAccountType(owner string) (ProviderAccountType, e
 	})
 
 	if err != nil {
-		if errors.Is(err, gitprovider.ErrNotFound) {
+		if errors.Is(err, gitprovider.ErrNotFound) || strings.Contains(err.Error(), "404 Group Not Found") {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:257]: It does but I couldnt find this specifice one.

File: pkg/gitproviders/provider_test.go [hunk 4]
```diff
@@ -30,6 +30,8 @@ var (
 	GithubUserTestName = "bot"
 	GitlabOrgTestName  = "weaveworks"
 	GitlabUserTestName = "bot"
+	GitHubProviderName = "github"
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:33]: yeah it probably does.

File: pkg/gitproviders/provider_test.go [hunk 5]
```diff
@@ -424,6 +447,85 @@ var _ = Describe("test personal repo exists", func() {
...
+	BeforeEach(func() {
+		client, recorder, err = getTestClientWithCassette("gitlab_repo_group_exists", GitLabProviderName)
+		Expect(err).NotTo(HaveOccurred())
+		gitProvider = defaultGitProvider{
+			domain:   gitlab.DefaultDomain,
+			provider: client,
+		}
+	})
+
+	It("succeed on validating repo existence", func() {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:505]: I was gonna do it in the pr with more tests so i could just get out the basic functionality but now it has to wait for the gitops stuff

File: pkg/services/app/add.go [hunk 6]
```diff
@@ -233,6 +233,10 @@ func (a *App) updateParametersIfNecessary(params AddParams) (AddParams, error) {
 		}
 	default:
 		// making sure url is in the correct format
+		if strings.Contains(params.Url, "git@gitlab.com:") {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/services/app/add.go:236]: I can move this to normalize. We dont do this check in there because it only applies to gitlab from what i saw.

File: pkg/services/app/add_test.go [hunk 7]
```diff
@@ -990,6 +990,63 @@ var _ = Describe("Test app hash", func() {
...
+
+		Describe("generates source manifest", func() {
+			It("creates GitRepository when source type is git", func() {
+				addParams.SourceType = wego.SourceTypeGit
+				err := appSrv.Add(addParams)
+				Expect(err).ShouldNot(HaveOccurred())
+
+				Expect(fluxClient.CreateSourceGitCallCount()).To(Equal(1))
+
+				name, url, branch, secretRef, namespace := fluxClient.CreateSourceGitArgsForCall(0)
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/services/app/add_test.go:1039]: I dont have the pr test in here because for some reason when I realized that the acceptance tests needed changes to check pull request I ignored it here too. I will add it.

File: test/acceptance/test/utils.go (hunk 1 above)
Comment (J-Thompson12) [reply test/acceptance/test/utils.go:327]: We are checking for this in the acceptance test.

File: pkg/gitproviders/provider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/gitproviders/provider.go:257]: Seems like a bug in `go-git-providers`. Might be worth documenting there.

File: test/acceptance/test/utils.go (hunk 1 above)
Comment (jpellizzari) [reply test/acceptance/test/utils.go:327]: Sorry not understanding. We check that something gets logged?

File: pkg/services/app/add_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/services/app/add_test.go:1039]: because in add unit test there is no extra code tested if gitlab is being used.

File: test/acceptance/test/utils.go (hunk 1 above)
Comment (J-Thompson12) [reply test/acceptance/test/utils.go:327]: ohh sorry I led you wrong. We use check the result of this function not the actual log. You are right though it does seem weird that we log it.

File: pkg/utils/common.go [hunk 8]
```diff
@@ -125,23 +125,28 @@ func ValidateNamespace(ns string) error {
 // user inputs that might be provided.
 func SanitizeRepoUrl(url string) string {
 	trimmed := ""
+	domain := "github"
+
+	if strings.Contains(url, "gitlab") {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/utils/common.go:130]: This will be a much bigger restructure. We cannot use normalizedRepoURL as a parameter because gitproviders imports common. We also sanitize the url in a bunch of random places.

[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/utils/common.go:130]: @J-Thompson12 Can you skip `utils` all together? Like, whatever is consuming this `SanitizeRepoURL` func can just use `NormalizedRepoURL`>

File: pkg/services/app/remove.go [hunk 9]
```diff
@@ -74,7 +75,12 @@ func (a *App) Remove(params RemoveParams) error {
 		return fmt.Errorf("failed to obtain config URL and branch: %w", err)
 	}
 
-	remover, err := a.cloneRepo(a.ConfigGit, cloneURL, branch, params.DryRun)
+	normalizedUrl, err := gitproviders.NewNormalizedRepoURL(cloneURL)
```
Comment (jpellizzari) [inline pkg/services/app/remove.go:78]: Having this so close to the `cloneRepo` method where it is consumed makes me wonder if we can get a `NormalizedRepoURL` somewhere earlier in the stack. Just a thought.

**Review by jpellizzari — COMMENTED** [review]
Looks good, just one one blocker around the protocol thing, then :+1:

File: pkg/gitproviders/provider.go [hunk 10]
```diff
@@ -593,50 +596,51 @@ type NormalizedRepoURL struct {
 	protocol   RepositoryURLProtocol
 }
 
-var sshPrefixRe = regexp.MustCompile(`git@(.*):(.*)/(.*)`)
+// normalizeRepoURLString accepts a url like git@github.com:someuser/podinfo.git and converts it into
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:599]: It was a godoc style comment but I figured it still had good information.

[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:599]: It doesnt need to be public anymore. That will hopefully force people to only use normalizedUrl

File: pkg/gitproviders/provider_test.go [hunk 11]
```diff
@@ -814,11 +917,11 @@ var _ = DescribeTable("NormalizedRepoURL", func(input string, expected expectedR
 		protocol: RepositoryURLProtocolSSH,
 	}),
 	Entry("gitlab https", "https://gitlab.com/someuser/podinfo.git", expectedRepoURL{
-		s:        "https://gitlab.com/someuser/podinfo.git",
+		s:        "ssh://git@gitlab.com/someuser/podinfo.git",
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:920]: normalizeRepoURLString was made to convert all of these to ssh. Now that everything uses normalizedURL these tests needed updated.

File: test/acceptance/test/add_tests.go [hunk 12]
```diff
@@ -204,7 +205,135 @@ var _ = Describe("Weave GitOps App Add Tests", func() {
...
+			Eventually(appRemoveOutput).Should(gbytes.Say("► Committing and pushing gitops updates for application"))
+			Eventually(appRemoveOutput).Should(gbytes.Say("► Pushing app changes to repository"))
+		})
+
+		By("And app should get deleted from the cluster", func() {
+			_ = waitForAppRemoval(appName, THIRTY_SECOND_TIMEOUT)
+		})
+	})
+
+	It("SmokeMy - Verify that gitops can deploy and remove a gitlab app that belongs in a subgroup", func() {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply test/acceptance/test/add_tests.go:275]: ahhh man thats what I put on there to run only the tests I want to run.

File: pkg/services/auth/auth.go [hunk 13]
```diff
@@ -96,6 +92,11 @@ func getGitProviderWithClients(
 	return provider, nil
 }
 
+type SecretName struct {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/services/auth/auth.go:95]: It just happened when I rebased jerrys changes. We had some conflicts on this and I just ended up using his stuff.

File: pkg/gitproviders/provider_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/gitproviders/provider_test.go:920]: > normalizeRepoURLString was made to convert all of these to ssh

That was changed in this PR, or it always worked that way?

If the former, should it do that?

[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:920]: It always did that. The original function this was testing was combined with sanitizeRepoUrl. The original one that tested this was normalizeRepoURL that would clean up the url for url parse to work. sanitizeRepoUrl was unchanged and took all urls and made them ssh

File: pkg/services/app/remove.go (hunk 9 above)
Comment (J-Thompson12) [reply pkg/services/app/remove.go:78]: I looked and we are not using url at any point before this.

File: .github/workflows/nightly.yaml [hunk 14]
```diff
@@ -100,6 +100,9 @@ jobs:
         GITHUB_ORG: weaveworks-gitops-test
         GITHUB_TOKEN: "${{ secrets.WEAVE_GITOPS_TEST_WEAVEWORKS_WEAVE_GITOPS_BOT_TOKEN }}"
         GITHUB_KEY: "${{ secrets.WEAVE_GITOPS_TEST_WEAVEWORKS_WEAVE_GITOPS_BOT_SSH_KEY }}"
+        GITLAB_ORG: weave-gitops
+        GITLAB_SUBGROUP: weabe-gitops-sub
```
Comment (jpellizzari) [inline .github/workflows/nightly.yaml:104]: `weabe` typo. Could be "correct" if the Gitlab group is also misspelled

**Review by jpellizzari — APPROVED** [review]

File: .github/workflows/nightly.yaml (hunk 14 above)
Comment (J-Thompson12) [reply .github/workflows/nightly.yaml]: Actually turns out they both were wrong. Unknown to me weave-gitops is a subgroup

File: pkg/apputils/utils.go [hunk 15]
```diff
@@ -160,7 +164,7 @@ func getGitClients(ctx context.Context, url, configUrl, namespace string, isHelm
 	}
 
 	if isExternalConfig {
-		configRepoClient, configRepoErr := authsvc.CreateGitClient(ctx, targetName, namespace, utils.SanitizeRepoUrl(configUrl))
+		configRepoClient, configRepoErr := authsvc.CreateGitClient(ctx, targetName, namespace, configUrl)
```
Comment (JamWils) [inline pkg/apputils/utils.go:167]: Are we sure this doesn't break anything?  If you follow `configUrl` as it is being passed down from different commands it can vary wildly.  I would double check to make sure this is the right change.  

💡 I wonder if the url should even be passed around, it might be worth a second pass refactoring to have most places take a `NormalizedUrl` instead of a `string`.

File: pkg/apputils/utils.go [hunk 16]
```diff
@@ -173,8 +177,8 @@ func getGitClients(ctx context.Context, url, configUrl, namespace string, isHelm
 	return appClient, configClient, authsvc.GetGitProvider(), nil
 }
 
-func getAuthService(ctx context.Context, providerUrl string) (auth.AuthService, error) {
-	gitProvider, providerErr := auth.GetGitProvider(ctx, providerUrl)
+func getAuthService(ctx context.Context, normalizedUrl gitproviders.NormalizedRepoURL) (auth.AuthService, error) {
```
Comment (JamWils) [inline pkg/apputils/utils.go:180]: :100: We need to start passing the `NormalizedRepoURL` instead of the original `string`.

**Review by JamWils — CHANGES_REQUESTED** [review]
Please address my blocking comments.  I know the scope of this story increased drastically please make sure that the PR description is updated to reflect that scope (e.g. subgroups is part of this work).  Overall, nice work.  I am concerned about one area where sanitize was removed,  I would double check how `configUrl` is being passed from various locations in the app, it might not be safe to just pass the `string` through.

There are a lot of edge cases and unit tests missing.  I would add these first and then tackle the recommended refactors.  It will be a lot easier to see if you broke existing functionality if you have those tests in place.

I know this wasn't a small PR and it was "scattered" in a lot of places.  Great work on the first iteration.

File: pkg/apputils/utils.go (hunk 15 above)
Comment (J-Thompson12) [reply pkg/apputils/utils.go:167]: This one has been normalized but i think you are correct. If everything passes a normalized URL you never have to wonder.

File: pkg/utils/common_test.go [hunk 17]
```diff
@@ -118,15 +117,3 @@ error occurred some error, retrying in 1s
 
 	})
 })
-
-var _ = DescribeTable("SanitizeRepoUrl", func(input string, expected string) {
-	result := SanitizeRepoUrl(input)
-	Expect(result).To(Equal(expected))
-},
-	Entry("git clone style", "git@github.com:someuser/podinfo.git", "ssh://git@github.com/someuser/podinfo.git"),
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/utils/common_test.go:126]: This code was combined with another function doing the same thing. So these tests were removed along with this function

File: pkg/gitproviders/provider.go [hunk 18]
```diff
@@ -564,6 +561,12 @@ func (p defaultGitProvider) waitUntilRepoCreated(ownerType ProviderAccountType,
 // The raw URL is assumed to be something like ssh://git@github.com/myorg/myrepo.git.
 // The common `git clone` variant of `git@github.com:myorg/myrepo.git` is not supported.
 func DetectGitProviderFromUrl(raw string) (GitProviderName, error) {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:563]: Ya it can now. Nice catch.

File: pkg/gitproviders/provider.go [hunk 19]
```diff
@@ -675,3 +679,19 @@ func (n NormalizedRepoURL) Provider() GitProviderName {
...
+	}
+
+	// Used to detect if a gitlab subgroup is used
+	if providerName == GitProviderGitLab {
+		if !strings.Contains(parts[len(parts)-3], "gitlab.com") {
+			return parts[len(parts)-3] + "/" + parts[len(parts)-2], nil
+		}
+	}
+
+	return parts[len(parts)-2], nil
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:696]: This was actually existing code. I can certainly change it.

File: pkg/git/gogit.go [hunk 20]
```diff
@@ -335,7 +334,7 @@ func (g *GoGit) GetRemoteUrl(dir string, remoteName string) (string, error) {
 		return "", fmt.Errorf("remote config in %s does not have an url", dir)
 	}
 
-	return utils.SanitizeRepoUrl(urls[0]), nil
+	return urls[0], nil
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/git/gogit.go:337]: Looking at the function we are not doing anything here that isnt already being tested. All the business logic is being handled by gogit

File: pkg/gitproviders/provider.go [hunk 21]
```diff
@@ -254,7 +253,7 @@ func (p defaultGitProvider) GetAccountType(owner string) (ProviderAccountType, e
 	})
 
 	if err != nil {
-		if errors.Is(err, gitprovider.ErrNotFound) {
+		if errors.Is(err, gitprovider.ErrNotFound) || strings.Contains(err.Error(), gitprovider.ErrGroupNotFound.Error()) {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:256]: Go git providers does not have the exact function that is returned. They have one that just says `404 group not found` and the full error has way more unnecessary info included in it.

File: pkg/gitproviders/provider_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:920]: we do have cases that test https and ssh missing. All the changes I made were just to the expected outcome.

File: .github/workflows/deploy.yml [hunk 22]
```diff
@@ -113,6 +118,10 @@ jobs:
         GITHUB_ORG: weaveworks-gitops-test
         GITHUB_TOKEN: "${{ secrets.WEAVE_GITOPS_TEST_WEAVEWORKS_WEAVE_GITOPS_BOT_TOKEN }}"
         GITHUB_KEY: "${{ secrets.WEAVE_GITOPS_TEST_WEAVEWORKS_WEAVE_GITOPS_BOT_SSH_KEY }}"
+        GITLAB_ORG: weave-gitops
+        GITLAB_SUBGROUP: weave-gitops-sub
```
Comment (J-Thompson12) [inline .github/workflows/deploy.yml:122]: I left this in there because I am working on a subgroup test but discovered there needs to be some work done to actually create a repo on a subgroup in the tests. I have written up an issue here https://app.zenhub.com/workspaces/weaveworks-foss-603cb6b1d4dff9000e242dc4/issues/weaveworks/weave-gitops/808

File: pkg/gitproviders/README.md [hunk 23]
```diff
@@ -38,7 +38,7 @@ var _ = Describe("create github repo", func() {
 	var err error
 	BeforeEach(func() {
 	    var client gitprovider.Client
-		client, recorder, err = getTestClientWithCassette("CASSETTE_ID")
+		client, recorder, err = getTestClientWithCassette("CASSETTE_ID", "provider name")
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/README.md:41]: done

File: pkg/gitproviders/provider.go [hunk 24]
```diff
@@ -593,50 +596,51 @@ type NormalizedRepoURL struct {
 	protocol   RepositoryURLProtocol
 }
 
-var sshPrefixRe = regexp.MustCompile(`git@(.*):(.*)/(.*)`)
+// normalizeRepoURLString accepts a url like git@github.com:someuser/podinfo.git and converts it into
+// a string like ssh://git@github.com/someuser/podinfo.git. This helps standardize the different
+// user inputs that might be provided.
+func normalizeRepoURLString(url string, providerName string) string {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go]: done

File: pkg/gitproviders/provider.go [hunk 25]
```diff
@@ -593,50 +596,51 @@ type NormalizedRepoURL struct {
...
-			return n
-		}
+	if trimmed != "" {
+		return fmt.Sprintf("ssh://git@%s.com/%s", providerName, trimmed)
 	}
 
 	return url
 }
 
 func NewNormalizedRepoURL(uri string) (NormalizedRepoURL, error) {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:623]: I have created a task around moving normalizedRepoURL and having it be called at the highest level so we dont ever pass  url as string.

[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:623]: https://app.zenhub.com/workspaces/weaveworks-foss-603cb6b1d4dff9000e242dc4/issues/weaveworks/weave-gitops/811

File: pkg/gitproviders/provider_test.go [hunk 26]
```diff
@@ -424,6 +447,85 @@ var _ = Describe("test personal repo exists", func() {
...
+	BeforeEach(func() {
+		client, recorder, err = getTestClientWithCassette("gitlab_repo_personal_exists", GitLabProviderName)
+		Expect(err).NotTo(HaveOccurred())
+		gitProvider = defaultGitProvider{
+			domain:   gitlab.DefaultDomain,
+			provider: client,
+		}
+	})
+
+	It("succeed on validating repo existence", func() {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:476]: added some

File: pkg/gitproviders/provider.go [hunk 27]
```diff
@@ -675,3 +679,19 @@ func (n NormalizedRepoURL) Provider() GitProviderName {
...
+
+func getOwnerFromUrl(url string, providerName GitProviderName) (string, error) {
+	parts := strings.Split(url, "/")
+	if len(parts) < 2 {
+		return "", fmt.Errorf("could not get owner from url %s", url)
+	}
+
+	// Used to detect if a gitlab subgroup is used
+	if providerName == GitProviderGitLab {
+		if !strings.Contains(parts[len(parts)-3], "gitlab.com") {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:688]: You are right. There are a couple spots in this file we use the string.

File: pkg/gitproviders/provider.go [hunk 28]
```diff
@@ -675,3 +679,19 @@ func (n NormalizedRepoURL) Provider() GitProviderName {
 func (n NormalizedRepoURL) Protocol() RepositoryURLProtocol {
 	return n.protocol
 }
+
+func getOwnerFromUrl(url string, providerName GitProviderName) (string, error) {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:680]: Sorry I need to get into the habit of making tests for functions that already exist.

File: pkg/gitproviders/provider.go [hunk 29]
```diff
@@ -675,3 +679,19 @@ func (n NormalizedRepoURL) Provider() GitProviderName {
...
+	parts := strings.Split(url, "/")
+	if len(parts) < 2 {
+		return "", fmt.Errorf("could not get owner from url %s", url)
+	}
+
+	// Used to detect if a gitlab subgroup is used
+	if providerName == GitProviderGitLab {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:687]: I get the comment on this. I feel like this is already such a small and manageable function to test that it doesn't make much sense to split it out into two functions. The logic wont ever be separate.

File: pkg/gitproviders/provider.go [hunk 30]
```diff
@@ -593,50 +596,51 @@ type NormalizedRepoURL struct {
...
+
+	normalized := normalizeRepoURLString(uri, string(providerName))
 
 	u, err := url.Parse(normalized)
 	if err != nil {
 		return NormalizedRepoURL{}, fmt.Errorf("could not create normalized repo URL %s: %w", uri, err)
 	}
 
-	owner, err := utils.GetOwnerFromUrl(normalized)
+	owner, err := getOwnerFromUrl(normalized, providerName)
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:636]: I can certainly do that.

File: pkg/gitproviders/provider_test.go [hunk 31]
```diff
@@ -800,11 +903,11 @@ var _ = DescribeTable("NormalizedRepoURL", func(input string, expected expectedR
 		protocol: RepositoryURLProtocolSSH,
 	}),
 	Entry("github https", "https://github.com/someuser/podinfo.git", expectedRepoURL{
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:915]: I don't think I understand what you are saying here. We are updating the github https to github ssh.

File: pkg/gitproviders/provider_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:515]: There are not any error cases to test here. The error cases are handled in other tests for GetAccountType

File: pkg/gitproviders/provider.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go]: fixed

File: pkg/gitproviders/provider.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go]: added test

File: pkg/gitproviders/provider.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:687]: updated to use url

File: pkg/gitproviders/provider.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go]: uses url now

File: pkg/gitproviders/provider.go [hunk 32]
```diff
@@ -593,50 +596,51 @@ type NormalizedRepoURL struct {
...
+	if strings.HasPrefix(url, sshPrefix) {
+		trimmed = strings.TrimPrefix(url, sshPrefix)
+	}
 
-		if len(matches) >= 3 {
-			provider := matches[1]
-			org := matches[2]
-			repo := matches[3]
-			n := fmt.Sprintf("ssh://git@%s/%s/%s", provider, org, repo)
+	httpsPrefix := fmt.Sprintf("https://%s.com/", providerName)
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go]: done

File: pkg/apputils/utils.go (hunk 15 above)
Comment (J-Thompson12) [reply pkg/apputils/utils.go]: this task needs to be handled in a different PR so we can do it right.

File: pkg/gitproviders/provider_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/gitproviders/provider_test.go:515]: I try to place comments below the recommended line and not above.  I was referring to the fact that `CreateRespository` only has a success test and not an error test.  I wasn't referring to the function `getAccounts` below.

File: pkg/gitproviders/provider_test.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/gitproviders/provider_test.go:930]: Disregard, I read that wrong.

File: pkg/gitproviders/provider_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/gitproviders/provider_test.go:945]: I see that now.  Thanks

File: pkg/gitproviders/provider.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/gitproviders/provider.go:687]: Yes, it's a lot cleaner now and easier to read.  To be honest if you aren't going to split it out and give it a name such as the example below then you might want to put the comment back in there.  Although your unit test does cover the case clearly (:100:) so someone could refer to that as well. 
```
isSubGroup := len(parts) > 3
if isSubGroup {
}
```

File: pkg/gitproviders/provider_test.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/gitproviders/provider_test.go:476]: Where?  I still only see the success test here.

File: pkg/gitproviders/README.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/gitproviders/README.md:41]: Ok, it would have taken 10 seconds to write: `Provider name is the git provider you are using such as `GitHub` or `GitLab`.

File: pkg/gitproviders/provider_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/gitproviders/provider_test.go:515]: Actually, I might be wrong about this, let me look again.

[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:515]: Yeah i know. Create repository only has a fail on getAccountType and CreateUserRepository which are both tested in this function. The only business logic done in this the function being tested is visibility and if its org or user.

File: pkg/gitproviders/provider_test.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:476]: Oh sorry I thought this was a different comment. Sorry there are way too many. This is the same as the above. I dont see a use case for the fail tests that are not covered in other tests

File: pkg/gitproviders/provider.go [hunk 33]
```diff
@@ -675,3 +677,18 @@ func (n NormalizedRepoURL) Provider() GitProviderName {
 func (n NormalizedRepoURL) Protocol() RepositoryURLProtocol {
 	return n.protocol
 }
+
+func getOwnerFromUrl(url url.URL, providerName GitProviderName) (string, error) {
+	parts := strings.Split(url.Path, "/")
+	if len(parts) < 2 {
```
Comment (JamWils) [inline pkg/gitproviders/provider.go:683]: Please write a test that passes in an empty url string and a string with only one "part".  Look at it from my perspective when I write a test, then write code, and then refactor (i.e. TDD).  I would pass in an empty and a second test with a partial url and it would actually panic because I don't have this if-statement in place.  I would then access the `parts[1]` and my code will panic because I don't have this check in place.  I would then add that check and both tests would pass.  We need to start capturing these sad path cases better and not just write happy path tests.

**Review by JamWils — CHANGES_REQUESTED** [review]
Just a couple minor comments on this review.  The code is a lot easier to read and understand, great work.  Please add some sad cases and I asked about a test for subgroups in provider_test.go.  In addition, the description on this PR still does not reflect the work that was done (i.e. subgroups), please update.

File: pkg/gitproviders/provider_test.go [hunk 34]
```diff
@@ -424,6 +457,85 @@ var _ = Describe("test personal repo exists", func() {
 
 })
 
+var _ = Describe("test gitlab project", func() {
```
[the remark this answers is not in the document]
Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:460]: Because this has the same issue with go-git-providers that the acceptance test had. This issue is written up here https://app.zenhub.com/workspaces/weaveworks-foss-603cb6b1d4dff9000e242dc4/issues/weaveworks/weave-gitops/808

File: pkg/gitproviders/provider.go (hunk 33 above)
Comment (J-Thompson12) [reply pkg/gitproviders/provider.go:685]: I have added sad paths for empty url and missing owner. I also added a check because I learned you can have subgroups in subgroups in gitlab. That isnt currently supported because that can get messy the way we have it.

File: pkg/gitproviders/provider_test.go (hunk 34 above)
[the remark this answers is not in the document]
Comment (JamWils) [reply pkg/gitproviders/provider_test.go:460]: Copy that, ➕ for including the link to the issue.  Thanks!!!

File: pkg/gitproviders/provider_test.go [hunk 35]
```diff
@@ -754,15 +866,41 @@ var _ = Describe("Test org deploy keys creation", func() {
...
+		Entry("github", "ssh://git@github.com/weaveworks/weave-gitops.git", GitProviderGitHub, "weaveworks"),
+		Entry("gitlab", "ssh://git@gitlab.com/weaveworks/weave-gitops.git", GitProviderGitLab, "weaveworks"),
+		Entry("gitlab with subgroup", "ssh://git@gitlab.com/weaveworks/sub_group/weave-gitops.git", GitProviderGitLab, "weaveworks/sub_group"),
 	)
 
+	DescribeTable("getOwnerFromUrl fail paths", func(normalizedUrl string, providerName GitProviderName, expected string) {
+		u, err := url.Parse(normalizedUrl)
+		Expect(err).NotTo(HaveOccurred())
+		_, err = getOwnerFromUrl(*u, providerName)
+		Expect(err).To(HaveOccurred())
```
Comment (JamWils) [inline pkg/gitproviders/provider_test.go:898]: Please check the actual error here.  We can jump on a chat as to why.

Comment (J-Thompson12) [reply pkg/gitproviders/provider_test.go:899]: done

**Review by JamWils — APPROVED** [review]
